python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py

```diff
@@ -0,0 +1,536 @@
+                # This instance has already been deleted.
+                continue
+
+            if instance.node_kind == NodeKind.HEAD:
+                # Not possible to delete head node.
+                continue
+
+            to_delete_instances_by_type[instance.node_type].append(instance)
+
+        scale_request = KuberayProvider.ScaleRequest(
+            desired_num_workers=num_workers_dict,
+            workers_to_delete=to_delete_instances_by_type,
+            worker_groups_to_clear_delete=worker_groups_without_pending_deletes,
+            worker_groups_with_pending_deletes=worker_groups_with_pending_deletes,
+        )
+
+        return scale_request
+
+    def _submit_scale_request(
+        self, scale_request: "KuberayProvider.ScaleRequest"
+    ) -> None:
+        """Submits a scale request to the Kuberay API server.
+
+        This method will convert the scale request to patches and submit the patches
+        to the Kuberay API server.
+
+        Args:
+            scale_request: The scale request.
+
+        Raises:
+            Exception: An exception is raised if the Kuberay API server returns an
+                error.
+        """
+        # Get the current ray cluster spec.
+        patch_payload = []
+
+        raycluster = self.ray_cluster
+
+        # Collect patches for replica counts.
+        for node_type, target_replicas in scale_request.desired_num_workers.items():
+            group_index = _worker_group_index(raycluster, node_type)
+            group_max_replicas = _worker_group_max_replicas(raycluster, group_index)
+            # Cap the replica count to maxReplicas.
+            if group_max_replicas is not None and group_max_replicas < target_replicas:
+                logger.warning(
+                    "Autoscaler attempted to create "
+                    + "more than maxReplicas pods of type {}.".format(node_type)
+                )
+                target_replicas = group_max_replicas
+            # Check if we need to change the target count.
+            if target_replicas == _worker_group_replicas(raycluster, group_index):
+                # No patch required.
+                continue
+            # Need to patch replica count. Format the patch and add it to the payload.
+            patch = worker_replica_patch(group_index, target_replicas)
+            patch_payload.append(patch)
+
+        # Maps node_type to nodes to delete for that group.
+        for (
+            node_type,
+            workers_to_delete_of_type,
+        ) in scale_request.workers_to_delete.items():
+            group_index = _worker_group_index(raycluster, node_type)
+            worker_ids_to_delete = [
+                worker.cloud_instance_id for worker in workers_to_delete_of_type
+            ]
+            patch = worker_delete_patch(group_index, worker_ids_to_delete)
+            patch_payload.append(patch)
+
+        # Clear the workersToDelete field for the worker groups that have been deleted.
```

Comment:
However, different from the original impl, I added the clear worker groups logic and no longer have the issue of side-effect mentioned here: https://github.com/ray-project/ray/blob/67a5b717b5e601543d3afd41387e5c79ebe30b7f/python/ray/autoscaler/_private/kuberay/node_provider.py#L372-L388